New Parsers for IOS-XR : show im dampening and show im dampening interface

gwoodwa1 opened the pull request:
New parsers created for IOS-XR : _show im dampening_ and _show im dampening interface_

Tested using the new method (as per below). Note there is a warning related to another parser which appears to be flagged whenever I run unit testing.

(sdk) gary@gary-MS-7B17:~/sdk/genieparser/src/genie/libs/parser/iosxr/tests$ python -m unittest discover
................................................................................................../home/gary/sdk/genieparser/src/genie/libs/parser/iosxr/show_controllers.py:78: FutureWarning: Possible nested set at position 74
  p2 = re.compile(r'^mac\=(?P<mac>[A-Fa-f0-9:]+) +vlan=(?P<vlan>\d+)'
/home/gary/sdk/genieparser/src/genie/libs/parser/iosxr/show_controllers.py:78: FutureWarning: Possible nested set at position 176
  p2 = re.compile(r'^mac\=(?P<mac>[A-Fa-f0-9:]+) +vlan=(?P<vlan>\d+)'
......................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................
----------------------------------------------------------------------
Ran 600 tests in 1.270s

OK

tahigash commented on src/genie/libs/parser/iosxr/tests/ShowImDampeningIntf/cli/equal/golden_output1_expected.py:
```diff
@@ -0,0 +1,14 @@
+expected_output = {
+            'interface': {
+                'GigabitEthernet0/2/0/0': {
+                    'currently_suppressed': 'no',
```
let's add 'interface_handle` here.
```
'GigabitEthernet0/2/0/0': {
    'interface_handle`: '0x080002c0',
```
corresponding output:
```
GigabitEthernet0/2/0/0 (0x080002c0)

gwoodwa1 replied:
Certainly, easy to add that into the schema.

tahigash replied:
confirmed the change

tahigash commented on src/genie/libs/parser/iosxr/tests/ShowImDampeningIntf/cli/equal/golden_output2_expected.py:
```diff
@@ -0,0 +1,31 @@
+                            'penalty': 1625,
+                            'protocol': {
+                                'ipv4': {
+                                    'protocol_capsulation': 'ipv4',
```
these ones has `protocol` level at line 9. so I believe we don't need prefix `procotol_`.
```
'protocol': {
    'ipv4': {
        'capsulation': 'ipv4',
        'penalty': 1615,
        'suppression': 'NO',
        'suppression_remaining_secs': 22,
        'underlying_state': 'Down',

gwoodwa1 replied:
Sure, I'll do that.. :)

tahigash replied:
```
                            'index': {
                                1: {
                                    'capsulation': 'ipv6',
                                    'penalty': 1625,
                                    'protocol': 'ipv6',
                                    'suppression': 'YES',
                                    'suppression_timer': 42,
                                    'underlying_state': 'Down'
                                },
```
can I ask you change from `suppression_timer` to `suppression_remaining_sec`? better to have keyword `remaining` and also unit `secs`.

gwoodwa1 replied:
Sorry I missed that from the first round, I'll make that change.

gwoodwa1 replied:
I've made that change and uploaded an amended version for that. Thanks.

tahigash replied:
confirmed the change

tahigash commented on src/genie/libs/parser/iosxr/tests/ShowImDampeningIntf/cli/equal/golden_output2_expected.py:
```diff
@@ -0,0 +1,31 @@
+                                'ipv4': {
+                                    'protocol_capsulation': 'ipv4',
+                                    'protocol_penalty': 1615,
+                                    'protocol_suppression': 'no',
```
let's take as is. no changing to lower case.
```
`suppression': 'NO'

gwoodwa1 replied:
Happy to do that, would you say the rule is leave all values "as is" and not to modify case or wording? I had another case example where a field just said "N" for No so I felt obliged to expand it to "No" so it was clear in terms of meaning. Just curious for anything I do in the future. Thank you.

tahigash replied:
@gwoodwa1 thanks for the comment. actually, we did modification in some parsers in the past. but it created some confusions why it's changed when user is looking for exact output. so we now try to respect the output as is. I think `Y` or `N` is still understandable/guessable. if we had `E`(Enabled) or `D`(Disable), it's a bit hard to guess. In that kind of case, we should consider modification. but usually, will take values as is as much as possible.

gwoodwa1 replied:
Cool, thanks for the clarification. That guidance helps alot for the future.

tahigash replied:
confirmed the change.

tahigash commented on src/genie/libs/parser/iosxr/tests/ShowImDampeningIntf/cli/equal/golden_output2_expected.py:
```diff
@@ -0,0 +1,31 @@
+                                    'protocol_penalty': 1615,
+                                    'protocol_suppression': 'no',
+                                    'protocol_suppression_timer': 22,
+                                    'protocol_underlying_state': 'down'
```
same. take as is. `Down`

gwoodwa1 replied:
Noted :)

tahigash commented on src/genie/libs/parser/iosxr/tests/ShowImDampeningIntf/cli/equal/golden_output1_expected.py:
```diff
@@ -0,0 +1,14 @@
+                    'penalty': 0,
+                    'reuse': 750,
+                    'suppress': 2000,
+                    'underlying_state': 'up'
```
take as is `Up`

gwoodwa1 replied:
Noted

tahigash replied:
confirmed the change

tahigash commented on src/genie/libs/parser/iosxr/tests/ShowImDampening/cli/equal/golden_output1_expected.py:
```diff
@@ -0,0 +1,41 @@
+                                'not_present': {
+                                    'capsulation': 'not_present',
+                                    'penalty': 2389,
+                                    'suppressed': 'yes'
```
take as is. `YES`

gwoodwa1 replied:
Noted, all these comments will be corrected in the next update. Thanks.

tahigash replied:
confirmed the change

tahigash reviewed (changes requested):
thank you for your contribution. left comments about schema.

gwoodwa1 commented on src/genie/libs/parser/iosxr/tests/ShowImDampening/cli/equal/golden_output1_expected.py:
```diff
@@ -0,0 +1,41 @@
+                    'interface': {
+                        'GigabitEthernet0/0/0/0': {
+                            'protocol': {
+                                'not_present': {
```
Makes sense but I just wondered if that was a hardcoded key of "1:" ? It looks a bit like a list so I wondered if there was special way of coding of it or just treat it as normal set of dict keys. I think I am right in thinking we couldn't have more than one index (as we have the interface as a unique key) so hard coding it exactly like that should be okay?

tahigash replied:
the number `1` will be incremented if we see same interface as multiple. the index number will start with `1` and then `2`, `3` and so on.

We usually don't have a list in the middle of dictionary. (only leaf has list as value). So, that's why we are doing `index` thing for this case.

gwoodwa1 replied:
Since Daniel's changes, the empty key values now report None instead. I assume this would negate the need for using a index here? See conversation with Daniel higher up where those changes are proposed. If we need to convert from None to index then please let me know. Thanks.

danielgraziano replied:
Hi @gwoodwa1, we still need the index instead of using protocol otherwise if there are two values with `None`, one would overwrite the other

tahigash replied:
confirmed the change

danielgraziano commented on src/genie/libs/parser/iosxr/show_dampening.py:
```diff
@@ -0,0 +1,241 @@
+            out = output
+            
+        result_dict = {}
+        result_dict.setdefault('interface', {})
```
I see where you are going with this. We need to move this inside the p1.match though. Reason being is we want the parser to return an empty dictionary if nothing parsed as that will raise a SchemaEmptyParserError.

gwoodwa1 replied:
Makes sense, I've moved that inside p1.match now. Thanks

danielgraziano commented on src/genie/libs/parser/iosxr/show_dampening.py:
```diff
@@ -0,0 +1,241 @@
+        
+        # GigabitEthernet0/0/0/1                                               2389 YES
+        # POS0/2/0/0                  <base>             ppp                      0 NO 
+        p1 = re.compile(r'^(?P<interface>.{1,26}) +(?P<prot>.{1,20}) +(?P<cap>.{1,10}) +(?P<pen>\d+) +(?P<sup>YES|NO)')
```
I feel this regex may be a little risky (unless you are confident about the spacing) and it seems to catch a lot of whitespace:
![image](https://user-images.githubusercontent.com/1134387/96029776-1db85600-0e29-11eb-9dc3-a1bd08cca6b3.png)

`^(?P<interface>\S+) +(?P<prot>\S+)? +(?P<cap>\S+)? +(?P<pen>\d+) +(?P<sup>YES|NO)` Might be better:
![image](https://user-images.githubusercontent.com/1134387/96029991-6bcd5980-0e29-11eb-818d-5bbf79cdcc1a.png)

gwoodwa1 replied:
Nice one Daniel, that suggestion is much better. I didn't know about the ? as an optional field. Thank you, I'll use that going forward.

danielgraziano commented on src/genie/libs/parser/iosxr/show_dampening.py:
```diff
@@ -0,0 +1,241 @@
+            else:
+                continue
+            
+            m = p1.match(line)
```
Please add an example on the line above for what this regex matches. Makes it easier to see at a glance

danielgraziano commented on src/genie/libs/parser/iosxr/show_dampening.py:
```diff
@@ -0,0 +1,241 @@
+            
+            if m:
+                group = m.groupdict()
+                interface=group['interface'].rstrip()
```
rstrip() wont be needed with new regex

danielgraziano commented on src/genie/libs/parser/iosxr/show_dampening.py:
```diff
@@ -0,0 +1,241 @@
+                        group[empty]="not_present"
+                protocol=group['prot'].rstrip()
+                
+                if result_dict['interface'].get(interface)==None:
+                    result_dict['interface'][interface]={}
+                    result_dict['interface'][interface]['protocol']={}
+                result_dict['interface'][interface]['protocol'][protocol]={}
+                result_dict['interface'][interface]['protocol'][protocol].update({'capsulation': group['cap']})  
+                result_dict['interface'][interface]['protocol'][protocol].update({'penalty': int(group['pen'])})
+                result_dict['interface'][interface]['protocol'][protocol].update({'suppressed': group['sup'].lower()})                                     
```
We can improve this like below as setdefault() will insert data if it doesnt exist but it also returns a reference to a point within a dictionary:
``` python
protocol_dict = result_dict.setdefault('interface', {}).setdefault(interface, {}).setdefault('protocol', {})
protocol_dict.update({'capsulation': group['cap']})
protocol_dict.update({'penalty': int(group['pen'])})
protocol_dict.update({'suppressed': group['sup'].lower()})

gwoodwa1 replied:
Hi Daniel, That works good. I didn't know from this that protocol_dict would also update result_dict as well. They look to be two separate dictionaries but clearly they are not. I am guessing this is because they share the same object in memory? However, whilst this has solved one problem it has created another in that empty fields e.g. capsulation has a value of "None"  which is busting my schema since it's not a string. I could just convert it to a string or change the schema to any() but is that okay? The same has happened in the protocol field as well where any empty fields are present. Data, below: {'interface': {'GigabitEthernet0/0/0/0': {'protocol': {None: {'capsulation': None, 'penalty': 629, 'suppressed': 'NO'}}}, 'GigabitEthernet0/0/0/1': {'protocol': {None: {'capsulation': None, 'penalty': 2389, 'suppressed': 'YES'}}}, 'POS0/2/0/0': {'protocol': {None: {'capsulation': None, 'penalty': 0, 'suppressed': 'NO'}, '<base>': {'capsulation': 'ppp', 'penalty': 0, 'suppressed': 'NO'}, 'ipv4': {'capsulation': 'ipcp', 'penalty': 0, 'suppressed': 'NO'}}}}}

gwoodwa1 replied:
For now, I've made the schema for capsulation to be Any() and it works fine. If this is not right then please let me know. Cheers.

danielgraziano replied:
Hi @gwoodwa1 for any optional items in the regex you can do:
``` python
if group['cap']:
  protocol_dict.update({'capsulation': group['cap']})
```
As we dont want to insert anything if it never existed

danielgraziano commented on src/genie/libs/parser/iosxr/show_dampening.py:
```diff
@@ -0,0 +1,241 @@
+    
+    cli_command = ['show im dampening interface {interface}']
+    
+    def cli(self, interface=None, output=None):
```
lets make interface mandatory. Below we can then remove the `if interface:` line

gwoodwa1 replied:
Certainly, no problem have done that now in the updated version.

danielgraziano commented on src/genie/libs/parser/iosxr/show_dampening.py:
```diff
@@ -0,0 +1,241 @@
+            out = output
+        
+        result_dict = {}
+        result_dict.setdefault('interface', {})
```
Same as the other parser. We want an empty dictionary unless the parser actually parses lines

gwoodwa1 replied:
Yeah, I've moved that inside the match p1 now instead. Cheers.

danielgraziano commented on src/genie/libs/parser/iosxr/show_dampening.py:
```diff
@@ -0,0 +1,241 @@
+        
+        # TenGigE 0/1/0/0 (0x01180020)
+        # GigabitEthernet0/2/0/0 (0x080002c0)
+        p1 = re.compile(r'^(?P<interface>[a-zA-Z].+[\/\d]+) \(')
```
Please match the whole line (As @tahigash mentioned we will need to save the handle)

gwoodwa1 replied:
Got that one, thanks :+1:

danielgraziano commented on src/genie/libs/parser/iosxr/show_dampening.py:
```diff
@@ -0,0 +1,241 @@
+        p6 = re.compile(r'suppress:\s+(?P<suppress>\d+) +max-suppress-time:\s(?P<max_suppress>\d+)')
+        
+        # ipv6           ipv6               1625  YES    42s  remaining        Down
+        p7 = re.compile(r'^(?P<prot>[a-z\d<>].{1,12}) +(?P<prot_cap>[a-z\d<>].{1,12})'
```
Same as the other parser. Are you confident about the spacing in this line? Otherwise we should make it match more strictly
![image](https://user-images.githubusercontent.com/1134387/96031135-f4002e80-0e2a-11eb-8939-8ead9d3a4b27.png)

gwoodwa1 replied:
Yeah, it's definitely looking better with your previous suggestion. Changed to the below:         
        p7 = re.compile(r'^(?P<prot>\S+)? +(?P<prot_cap>\S+)? +(?P<prot_pen>\d+) +(?P<prot_sup>YES|NO)'
                         ' +(?P<prot_sup_time>\d+). +remaining\s+(?P<prot_state>\S+)')

danielgraziano commented on src/genie/libs/parser/iosxr/show_dampening.py:
```diff
@@ -0,0 +1,241 @@
+            else:
+                continue
+
+            m = p1.match(line)
```
Please add comment above all regex to display line that can be matched

gwoodwa1 replied:
Hi Daniel, Each re.compile line has a raw output line with a # comment to show what it is matching. Am I missing something? Could you point me at a specific example as I think I've done it already but maybe I am not seeing it? I've even shown two examples in some cases where the raw output varies e.g. underlying-state which has two ways of being outputted. underlying-state:  Up and Underlying state: Down. Thanks

danielgraziano replied:
Hi @gwoodwa1 we just like to have the same comment above both the re.compile and the .match(). Otherwise looking at this line I need to scroll up to the p1 definition to see what it matches. So do something like:

``` python
# TenGigE 0/1/0/0 (0x01180020)
# GigabitEthernet0/2/0/0 (0x080002c0)
p1 = re.compile(r'^(?P<interface>[a-zA-Z].+[\/\d]+) \(')

(snip)

# TenGigE 0/1/0/0 (0x01180020)
# GigabitEthernet0/2/0/0 (0x080002c0)
m = p1.match(line)

danielgraziano commented on src/genie/libs/parser/iosxr/show_dampening.py:
```diff
@@ -0,0 +1,241 @@
+            if m:
+                group = m.groupdict()
+                interface=group['interface']
+                if result_dict['interface'].get(interface)==None:
```
Same as other parser, we can improve the code flow

``` python
interface_dict = result_dict.setdefault('interface', {}).setdefault(interface, {})

# Then we can use the interface_dict reference in other matches below as well

gwoodwa1 replied:
Changed in the newer version. Thanks.

gwoodwa1 commented:
I believe that I have I have addressed all the feedback and comments in the latest version. However, there was some significant amount of changes to the schema and code so please let me know if you think I missed anything. I've re-run the unit test successfully see below, Thank you for all your feedback as this is helping me improve the quality of these submissions.


(sdk) gary@gary-MS-7B17:~/sdk/genieparser/src/genie/libs/parser/iosxr/tests$ python -m unittest discover
................................................................................................../home/gary/sdk/genieparser/src/genie/libs/parser/iosxr/show_controllers.py:78: FutureWarning: Possible nested set at position 74
  p2 = re.compile(r'^mac\=(?P<mac>[A-Fa-f0-9:]+) +vlan=(?P<vlan>\d+)'
/home/gary/sdk/genieparser/src/genie/libs/parser/iosxr/show_controllers.py:78: FutureWarning: Possible nested set at position 176
  p2 = re.compile(r'^mac\=(?P<mac>[A-Fa-f0-9:]+) +vlan=(?P<vlan>\d+)'
......................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................
----------------------------------------------------------------------
Ran 600 tests in 1.252s

OK

tahigash reviewed (changes requested):
only 1 comment left from me.

tahigash reviewed (approved):
schema looks good now.

danielgraziano commented on src/genie/libs/parser/iosxr/show_dampening.py:
```diff
@@ -0,0 +1,257 @@
+
+class ShowImDampeningSchema(MetaParser):
+    schema = {
+                'interface': {
```
Please fix the indentation for this schema

gwoodwa1 replied:
Fixed

danielgraziano commented on src/genie/libs/parser/iosxr/show_dampening.py:
```diff
@@ -0,0 +1,257 @@
+    schema = {
+        'interface': {
+            Any(): {
+            'dampening_status': str,
```
Please fix the indentation for this schema

gwoodwa1 replied:
Fixed

danielgraziano commented on src/genie/libs/parser/iosxr/tests/ShowImDampeningIntf/cli/equal/golden_output1_expected.py:
```diff
@@ -0,0 +1,15 @@
+expected_output = {
+                    'interface': {
```
Please fix indentation

gwoodwa1 replied:
Fixed

danielgraziano commented on src/genie/libs/parser/iosxr/tests/ShowImDampeningIntf/cli/equal/golden_output2_expected.py:
```diff
@@ -0,0 +1,34 @@
+expected_output = {
+                    'interface': {
```
Please fix indentation

gwoodwa1 replied:
Fixed

danielgraziano commented on src/genie/libs/parser/iosxr/tests/ShowImDampeningIntf/cli/equal/golden_output3_expected.py:
```diff
@@ -0,0 +1,8 @@
+
+expected_output = {
+                    'interface': {
```
Please fix indentation

gwoodwa1 replied:
Fixed

danielgraziano commented:
Hi @gwoodwa1, It looks great! Just fix the indentations that I left comments on and then we can go for merge! Thanks for the contribution!

gwoodwa1 commented:
Fixed those indentation issues, I am going to leave auto PEP-8 on in future so those get automatically fixed. Hopefully, that is it. Thanks for your all your feedback and support.
